rep-2007.rst

```diff
@@ -0,0 +1,246 @@
+Author: Audrow Nash <audrow@openrobotics.org>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 28-Jan-2020
+Post-History: 28-Jan-2020
+
+
+Abstract
+========
+
+This REP proposes an extension to the creation of Publishers and Subscriptions in `rclcpp` to make it easier to convert between ROS 2 interfaces and custom C++ data structures.
+
+
+Motivation
+==========
+
+The primary reason for this change is to improve the developer's experience working with ROS 2.
+Currently, developers often write code to convert a ROS interface into another custom data structure for use in their programs.
+This can be trivial, in the case accessing the ``data`` field in ``std_msgs::msg::String``;
+or more involved such when converting OpenCV's ``cv::Map`` to ROS's ``sensor_msgs/msg/Image`` type [1]_.
+Such conversions are additional work for the programmer and are potential sources of errors.
+
+An additional reason for this change is performance.
+This interface will allow us to define methods for serializing directly to the user requested type, and/or using that type in intra-process communication without ever converting it.
+Whereas without this feature, even to send a ``cv::Mat`` from one publisher to a subscription in the same process would require converting it to a ``sensor_msgs::msg::Image`` first and then back again to cv::Mat.
+
+
+Terminology
+===========
+
+.. glossary::
+     Custom type (in code, ``CustomType``)
+       A data structure that *is not* a ROS 2 interface, such as ``std::string`` or ``cv::Mat``.
+     ROS type (in code, ``RosType``)
+       A data structure that *is* a ROS 2 interface, such as ``std_msgs::msg::String`` or ``sensor_msgs::msg::Image``.
+
+
+Specification
+=============
+
+There are different tiers of expressivity in the syntax for type adaptation, including both explicit to implicit forms.
+The explicit forms require the custom type and the ROS type to be specified; 
+while the implicit form makes an assumption about the ROS type based on the custom type.
+The explicit forms may be used to clearly communicate what is happening, which could be useful in examples or tutorials, or may be used depending on the preferences of the developer.
+While the implicit, and most concise, form will primarily be used for its convenience (although it lacks some of the functionality of the explicit forms).
+
+Below are examples of different syntaxes for type adaptation.
+In these examples and using the terminology above, ``std::string`` will be considered the custom type and the ROS interface ``std_msgs::msg::String`` will be considered the ROS type.
+Note that, for every publisher created in the examples below, a ``std::string`` type variable (the custom type) would be adapted to be published on the ROS network as a ``std_msgs::msg::String`` (the ROS type).
+
+.. code-block:: cpp
+
+   // Explicit Form 1: Most explicit
+   using MsgT = rclcpp::AdaptType<std::string>::to<std_msgs::msg::String>;
```

Comment:
One thing I'm wondering about: is the type adapted (ie: changed in some way) or is machinery/infrastructure added around the type which allows it to be used in a way it otherwise could not be (ie: directly published and subscribed)?

I'm not a native speaker (but I do have a CS background, so am familiar with the Facade and Proxy patterns fi), but adapting the type sounds like it's going to be changed to allow it to be used in-place of a `std_msgs::msg::String`. Adding functionality around it would not seem to change the type per se.

Reply:
Yeah, I guess a more accurate way of thinking about it is that we're creating an "adapter" which allows you to "fit" a new type, like `std::string`, into the interface intended originally for `std_msgs::msg::String`.

So in that sense, perhaps `AdaptType` (which is an action and indicates we're doing something to the target type as you said) isn't very correct.

We could go back to `TypeAdapter<std::string>` which would read "Type adapter for type std::string" or "Type adapter of type std::string".

So in that language, you'd say, "A publisher takes either a ROS Message Type or a Type Adapter for a ROS Message Type". The implicit style from ROS 1 where you pass just the custom type would implicitly be upgraded to a Type Adapter for you.  Does that make more sense?